irods/at_client_exit.py

```diff
@@ -0,0 +1,43 @@
+import atexit
+import enum
+import logging
+
+logger = logging.getLogger(__name__)
+
+class priority(enum.Enum):
+    # Integer value determines relative order of execution among the stages.
+    BEFORE_PRC_CLEANUP = 10
+    DURING_PRC_CLEANUP = 20
+    AFTER_PRC_CLEANUP = 30
+    def __lt__(self,other):
+        return self.value < other.value
+
+
+# So that we can access priority.ENUM_NAME as simply ENUM_NAME at module level.
+for _ in priority.__members__:
+    exec('{}=priority.{}'.format(_,_))
+
+_cleanup_functions = {BEFORE_PRC_CLEANUP:[], DURING_PRC_CLEANUP:[], AFTER_PRC_CLEANUP:[]}
+
+initialized = False
+
+def register(priority_for_execution, function):
+    global initialized
+    if not initialized:
+        initialized = True
+        atexit.register(call_cleanup_functions)
+    array = _cleanup_functions[priority_for_execution]
+    if function not in array:
+        array.append(function)
+
+def call_cleanup_functions():
+   ordered_funclists = sorted((pri,fl) for (pri,fl) in _cleanup_functions.items())
+   # Run each cleanup stage.
+   for priority,function_list in ordered_funclists:
+       # Within each cleanup stage, run all registered functions last-in, first-out. (Per atexit conventions.)
+       for function in reversed(function_list):
```

Comment:
One of the cool thing about most of the  functions that transform data structures in Python3 is that they return iterators rather than doing a transformed copy of the original data structure So, we don't have to fear duplicating the data or doubling the storage, in effect, contained by a large list. `reversed(L) `thus returns an iterator to the sequence L, that iterates backward through the list.  `dict.keys` is another such example.

Reply:
I think perhaps we can resolve this one?